File: examples/streaming_decompression.c
/*
 * Copyright (c) 2016-present, Yann Collet, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under both the BSD-style license (found in the
 * LICENSE file in the root directory of this source tree) and the GPLv2 (found
 * in the COPYING file in the root directory of this source tree).
 * You may select, at your option, one of the above-listed licenses.
 */


#include <stdlib.h>    // malloc, exit
#include <stdio.h>     // fprintf, perror, feof
#include <string.h>    // strerror
#include <errno.h>     // errno
#include <zstd.h>      // presumes zstd library is installed
#include "utils.h"

static void decompressFile_orDie(const char* fname)
{
    FILE* const fin  = fopen_orDie(fname, "rb");
    size_t const buffInSize = ZSTD_DStreamInSize();
    void*  const buffIn  = malloc_orDie(buffInSize);
    FILE* const fout = stdout;
    size_t const buffOutSize = ZSTD_DStreamOutSize();  /* Guarantee to successfully flush at least one complete compressed block in all circumstances. */
    void*  const buffOut = malloc_orDie(buffOutSize);

    ZSTD_DStream* const dstream = ZSTD_createDStream();
    if (dstream==NULL) { fprintf(stderr, "ZSTD_createDStream() error \n"); exit(10); }

    /* In more complex scenarios, a file may consist of multiple appended frames (ex : pzstd).
    *  The following example decompresses only the first frame.
    *  It is compatible with other provided streaming examples */
    size_t const initResult = ZSTD_initDStream(dstream);
    if (ZSTD_isError(initResult)) { fprintf(stderr, "ZSTD_initDStream() error : %s \n", ZSTD_getErrorName(initResult)); exit(11); }
    size_t read, toRead = initResult;
    while ( (read = fread_orDie(buffIn, toRead, fin)) ) {
        ZSTD_inBuffer input = { buffIn, read, 0 };
        while (input.pos < input.size) {
            ZSTD_outBuffer output = { buffOut, buffOutSize, 0 };
            toRead = ZSTD_decompressStream(dstream, &output , &input);  /* toRead : size of next compressed block */
            if (ZSTD_isError(toRead)) { fprintf(stderr, "ZSTD_decompressStream() error : %s \n", ZSTD_getErrorName(toRead)); exit(12); }
            fwrite_orDie(buffOut, output.pos, fout);
        }
    }

    ZSTD_freeDStream(dstream);
    fclose_orDie(fin);
    fclose_orDie(fout);
    free(buffIn);
    free(buffOut);
}


int main(int argc, const char** argv)
{
    const char* const exeName = argv[0];

    if (argc!=2) {
        fprintf(stderr, "wrong arguments\n");
        fprintf(stderr, "usage:\n");
        fprintf(stderr, "%s FILE\n", exeName);
        return 1;
    }

    const char* const inFilename = argv[1];

    decompressFile_orDie(inFilename);
    return 0;
}

File: examples/streaming_decompression.zig
const std = @import("std");
const zstd = @import("zstd");
const common = @import("common.zig");

pub fn main() !void {
    const allocator = std.heap.c_allocator;
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if (args.len != 2) {
        std.debug.print("wrong arguments\nusage:\n{s} FILE\n", .{args[0]});
        return;
    }

    const input_filename = args[1];

    const fin = try std.fs.cwd().openFile(input_filename, .{});
    defer fin.close();

    // Write to stdout (std.io.getStdOut() is missing in 0.15?, using debug print as fallback)
    // const stdout = std.io.getStdOut().writer();

    const buffInSize = zstd.c.ZSTD_DStreamInSize();
    const buffIn = try allocator.alloc(u8, buffInSize);
    defer allocator.free(buffIn);

    const buffOutSize = zstd.c.ZSTD_DStreamOutSize();
    const buffOut = try allocator.alloc(u8, buffOutSize);
    defer allocator.free(buffOut);

    const dctx = zstd.c.ZSTD_createDCtx();
    if (dctx == null) return error.ZstdCreateDCtxFailed;
    defer _ = zstd.c.ZSTD_freeDCtx(dctx);

    // var r = fin.reader();

    while (true) {
        const read = try fin.read(buffIn);
        if (read == 0) break;

        var input = zstd.c.ZSTD_inBuffer{
            .src = buffIn.ptr,
            .size = read,
            .pos = 0,
        };

        while (input.pos < input.size) {
            var output = zstd.c.ZSTD_outBuffer{
                .dst = buffOut.ptr,
                .size = buffOutSize,
                .pos = 0,
            };

            const ret = zstd.c.ZSTD_decompressStream(dctx, &output, &input);
            if (zstd.c.ZSTD_isError(ret) != 0) {
                std.debug.print("Decompression error: {s}\n", .{zstd.c.ZSTD_getErrorName(ret)});
                return error.DecompressionFailed;
            }

            // try stdout.writeAll(buffOut[0..output.pos]);
            std.debug.print("{s}", .{buffOut[0..output.pos]});
        }
    }
}



